README.md
# IoTaaP HAPI - Hardware API Library
> Hardware to Cloud Application Programming Interface

[![Build Status](https://jenkins.iotaap.io/buildStatus/icon?job=iotaap-hapi%2Fmaster)](https://jenkins.iotaap.io/job/iotaap-hapi/job/master/)

 ## Switching to IoTaaP OS

Please be aware that this library is deprecated and we have introduced IoTaaP OS from 2021-01-29, and for any new project
we recommend using IoTaaP OS, although IoTaaP HAPI will still be available.
[**IoTaaP OS is on GitHub**](https://github.com/iotaap/iotaap-os)

## Usage

Include `IoTaaP_HAPI.h` in your project.

## Usage example

```
#include "IoTaaP_HAPI.h"

IoTaaP_HAPI iotaap_hapi;
```

## Hardware

Use official [IoTaaP Hardware modules](https://www.iotaap.io/)

* [IoTaaP Magnolia](https://www.iotaap.io/platform/)

## Tutorials

Follow [official tutorials](https://docs.iotaap.io/iotaap-tutorials/) for even faster IoT development process!

## Documentation

Official [documentation](https://docs.iotaap.io)

## Community

Join [IoTaaP community](https://community.iotaap.io) today!

## Authors

* [IoTaaP Team](https://www.iotaap.io) – contact@iotaap.io
* IoTaaP Community

## License
By contributing, you agree that your contributions will be licensed under Apache License 2.0, see [LICENSE](./LICENSE.md).

CHANGELOG.md
# Change Log

## [1.1.7] - 2021-01-29
 
### Added

### Changed

**!!!Breaking changes!!!**

New certificate is added to the project and implemented directly to the library. Now the same Root CA certificate can be 
used both for MQTTS and OTA Updates. 

In order to use previous version of IoTaaP HAPI Library you can select it using platformio.ini file by adding:
> lib_deps=
    iotaap/IoTaaP HAPI @ 1.1.6

Please be aware that this library is deprecated and we have introduced IoTaaP OS from 2021-01-29, and for any new project
we recommend using IoTaaP OS, although IoTaaP HAPI will still be available.
[**IoTaaP OS is on GitHub**](https://github.com/iotaap/iotaap-os)
   
### Fixed

## [1.1.6] - 2021-01-27
 
### Added

### Changed

License
   
### Fixed

Temporary fix in example for: https://community.platformio.org/t/adafruit-gfx-lib-will-not-build-any-more-pio-5/15776

## [1.1.5] - 2020-04-04
 
### Added

Added functions to enable or disable updates
 
### Changed

Made checkUpdate() public
   
### Fixed

## [1.1.4] - 2020-03-30
 
### Added

Added travis testing and basic example
 
### Changed

Updated IoTaaP Core version dependency
   
### Fixed

Fixed wrong certificate path

## [1.1.3] - 2020-03-30
 
### Added

Added detection if device is standalone or in a group
 
### Changed

Updated IoTaaP Core version dependency
   
### Fixed

Fixed buggy fw version variable
Fixed device publish function with missing topic

## [0.1.3] - 2020-03-30
 
### Added

OTA Upgrade
 
### Changed
   
### Fixed

## [0.1.0] - 2020-03-08
 
### Added

First stable version. Status publishing, publish and device publish working.
Added new IoTaaP Cloud certificate.
Added some future functions templates.
 
### Changed
   
### Fixed
     
## [0.0.1] - 2019-09-15
  
Initial version
 
### Added

Initial documentation
 
### Changed
   
### Fixed

src/IoTaaP_HAPI.h
/*
 * IoTaaP HAPI - Hardware API Library
 * Connectivity platform for rapid development
 * 
 * IoTaaP Team <contact@iotaap.io>
 * https://www.iotaap.io
 * 
 */

#ifndef __IOTAAP_HAPI_H__
#define __IOTAAP_HAPI_H__

#warning "IoTaaP HAPI library is Deprecated - Use IoTaaP OS"

#define HAPI_VERSION "1.1.7"
#define DEVICE_STATUS_PERIOD 500
#define DEVICE_OTA_CHECK_PERIOD 30000

#include <Arduino.h>
#include <IoTaaP.h>
#include <WiFi.h>
#include <WiFiMulti.h>
#include <HTTPClient.h>
#include <HTTPUpdate.h>
#include <time.h>

class IoTaaP_HAPI
{
public:
    IoTaaP_HAPI(const char *fwVersion);
    int configure(const char *deviceID, const char *deviceToken, const char *mqttServer, const char *mqttUsername, const char *mqttPassword, MQTT_CALLBACK_SIGNATURE, const char *groupID = '\0', const char *groupToken = '\0');
    int devicePublish(const char *payload, const char *uTopic);
    int publish(const char *payload, const char *topic);
    int subscribe(const char *topic);
    int unsubscribe(const char *topic);
    void apiLoop(bool sendStates = false);
    void callbackInnerFunction(char *topic, byte *message, unsigned int length);
    void enableUpdates();
    void disableUpdates();
    void checkUpdate();

private:
    IoTaaP iotaapCore;
    int connectToCloud(const char *server, const char *user, const char *password, MQTT_CALLBACK_SIGNATURE, const char *clientID = '\0');
    int publishStatus();
    void setClock();
    void otaUpdate();
    WiFiClientSecure _client;
    HTTPClient _httpClient;
    bool _updatesEnabled;
    const char *_deviceID;
    const char *_deviceToken;
    const char *_mqttServer;
    const char *_mqttUsername;
    const char *_mqttPassword;
    const char *_groupID;
    const char *_groupToken;
    const char *_fwVersion;
    unsigned long int _uptime;
    unsigned long int _otaUpdateNow;  // Detect passed time between update check
    unsigned long int _otaUpdatePrev; // Detect passed time between update check
    unsigned long int _sentMessages;
    unsigned long int _receivedMessages;
    unsigned long int _disconnects;
    unsigned long int _uptimePrev;
    unsigned long int _uptimeCur;
};

#endif

src/certificate.h
/*
 * IoTaaP Cloud Certificate
 * 
 * IoTaaP Team <contact@iotaap.io>
 * https://www.iotaap.io
 * 
 */

#ifndef __IOTAAP_CLOUD_CERTIFICATE_H__
#define __IOTAAP_CLOUD_CERTIFICATE_H__

const char *iotaap_mqtts_certificate =
    "-----BEGIN CERTIFICATE-----\n"
    "MIIDlTCCAn2gAwIBAgIUd6oLZtAfYIjLOg2908R3hxQZ3lUwDQYJKoZIhvcNAQEL\n"
    "BQAwWjELMAkGA1UEBhMCSFIxEDAOBgNVBAgMB0Nyb2F0aWExDzANBgNVBAcMBlph\n"
    "Z3JlYjEPMA0GA1UECgwGSW9UYWFQMRcwFQYDVQQDDA5tcXR0LmlvdGFhcC5pbzAe\n"
    "Fw0yMDAzMDgxMTA3MzhaFw0yNTAzMDcxMTA3MzhaMFoxCzAJBgNVBAYTAkhSMRAw\n"
    "DgYDVQQIDAdDcm9hdGlhMQ8wDQYDVQQHDAZaYWdyZWIxDzANBgNVBAoMBklvVGFh\n"
    "UDEXMBUGA1UEAwwObXF0dC5pb3RhYXAuaW8wggEiMA0GCSqGSIb3DQEBAQUAA4IB\n"
    "DwAwggEKAoIBAQCtKghvPBU2xnOXsx0W/5y1ew8AYuws5cSkzukgzTkQI0XTO0n5\n"
    "IT8CQD86Qi5/Q5qQtw0BQMUi5DEAPv/pHI0KuAdOaWYqlzA7AguP/4ubHvyJCNve\n"
    "Id6584YGLLw3YRSHAAFuMSwUd09NcLi21CwwSnH8K6GYhSDzYC/N9E6vzlpmSb8f\n"
    "48FeFkawHMrBgA/OFN7ffmIXe3QoCl4Ui3czk/rG3GN4J0rwpt16H8nvbGymIjBp\n"
    "9cmya8US9lFDbpuCDaKTlH7Dla6ibMkDpOA8JnwukJEwiaEJR5wN9n3NwY/JEVXa\n"
    "SXyEtQTKAqPoZj1y3CSmq/Y8FfPz1IkwPzMNAgMBAAGjUzBRMB0GA1UdDgQWBBSZ\n"
    "m4E30lB6RHa1slkKEmFUMbZFWDAfBgNVHSMEGDAWgBSZm4E30lB6RHa1slkKEmFU\n"
    "MbZFWDAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQB7fWkOQ1Uj\n"
    "4/W4LLkFCHa+OYZIy+t3r8bgYjKoCqH8+DlXZWpwThjn8a+heTeYCknbh/D8270k\n"
    "edzBhUjK06qoDWpXOI4eL3+7vffLS+AEanAi/XuiqYi7spoMSz2JwXFjc/94HpYf\n"
    "WC4fcelqIdQl0qR4dnRH9otKfz4KeqhgVcjL0nyPpt2Q9UndN0nT043bthW5kk35\n"
    "f2eoPel0okXFNmcvusS1pEvg8ddfM6GkJTYZr5aj5ZjitRE0ZwGBayE76JkLrSlZ\n"
    "fbPnrsnBM5OA7ytgR7LyFB4bHaJCM81NAYDQGG9hcsiaZeCEOQHIR97xd32a52Tm\n"
    "QySj9+LAkqon\n"
    "-----END CERTIFICATE-----\n";

#endif

src/certificate_new.h
/*
 * IoTaaP Cloud Certificate
 * 
 * IoTaaP Team <contact@iotaap.io>
 * https://www.iotaap.io
 * 
 */

#ifndef __IOTAAP_CLOUD_CERTIFICATE_H__
#define __IOTAAP_CLOUD_CERTIFICATE_H__

const char *iotaap_certificate =
    "-----BEGIN CERTIFICATE-----\n"
    "MIIGEzCCA/ugAwIBAgIQfVtRJrR2uhHbdBYLvFMNpzANBgkqhkiG9w0BAQwFADCB\n"
    "iDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCk5ldyBKZXJzZXkxFDASBgNVBAcTC0pl\n"
    "cnNleSBDaXR5MR4wHAYDVQQKExVUaGUgVVNFUlRSVVNUIE5ldHdvcmsxLjAsBgNV\n"
    "BAMTJVVTRVJUcnVzdCBSU0EgQ2VydGlmaWNhdGlvbiBBdXRob3JpdHkwHhcNMTgx\n"
    "MTAyMDAwMDAwWhcNMzAxMjMxMjM1OTU5WjCBjzELMAkGA1UEBhMCR0IxGzAZBgNV\n"
    "BAgTEkdyZWF0ZXIgTWFuY2hlc3RlcjEQMA4GA1UEBxMHU2FsZm9yZDEYMBYGA1UE\n"
    "ChMPU2VjdGlnbyBMaW1pdGVkMTcwNQYDVQQDEy5TZWN0aWdvIFJTQSBEb21haW4g\n"
    "VmFsaWRhdGlvbiBTZWN1cmUgU2VydmVyIENBMIIBIjANBgkqhkiG9w0BAQEFAAOC\n"
    "AQ8AMIIBCgKCAQEA1nMz1tc8INAA0hdFuNY+B6I/x0HuMjDJsGz99J/LEpgPLT+N\n"
    "TQEMgg8Xf2Iu6bhIefsWg06t1zIlk7cHv7lQP6lMw0Aq6Tn/2YHKHxYyQdqAJrkj\n"
    "eocgHuP/IJo8lURvh3UGkEC0MpMWCRAIIz7S3YcPb11RFGoKacVPAXJpz9OTTG0E\n"
    "oKMbgn6xmrntxZ7FN3ifmgg0+1YuWMQJDgZkW7w33PGfKGioVrCSo1yfu4iYCBsk\n"
    "Haswha6vsC6eep3BwEIc4gLw6uBK0u+QDrTBQBbwb4VCSmT3pDCg/r8uoydajotY\n"
    "uK3DGReEY+1vVv2Dy2A0xHS+5p3b4eTlygxfFQIDAQABo4IBbjCCAWowHwYDVR0j\n"
    "BBgwFoAUU3m/WqorSs9UgOHYm8Cd8rIDZsswHQYDVR0OBBYEFI2MXsRUrYrhd+mb\n"
    "+ZsF4bgBjWHhMA4GA1UdDwEB/wQEAwIBhjASBgNVHRMBAf8ECDAGAQH/AgEAMB0G\n"
    "A1UdJQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAbBgNVHSAEFDASMAYGBFUdIAAw\n"
    "CAYGZ4EMAQIBMFAGA1UdHwRJMEcwRaBDoEGGP2h0dHA6Ly9jcmwudXNlcnRydXN0\n"
    "LmNvbS9VU0VSVHJ1c3RSU0FDZXJ0aWZpY2F0aW9uQXV0aG9yaXR5LmNybDB2Bggr\n"
    "BgEFBQcBAQRqMGgwPwYIKwYBBQUHMAKGM2h0dHA6Ly9jcnQudXNlcnRydXN0LmNv\n"
    "bS9VU0VSVHJ1c3RSU0FBZGRUcnVzdENBLmNydDAlBggrBgEFBQcwAYYZaHR0cDov\n"
    "L29jc3AudXNlcnRydXN0LmNvbTANBgkqhkiG9w0BAQwFAAOCAgEAMr9hvQ5Iw0/H\n"
    "ukdN+Jx4GQHcEx2Ab/zDcLRSmjEzmldS+zGea6TvVKqJjUAXaPgREHzSyrHxVYbH\n"
    "7rM2kYb2OVG/Rr8PoLq0935JxCo2F57kaDl6r5ROVm+yezu/Coa9zcV3HAO4OLGi\n"
    "H19+24rcRki2aArPsrW04jTkZ6k4Zgle0rj8nSg6F0AnwnJOKf0hPHzPE/uWLMUx\n"
    "RP0T7dWbqWlod3zu4f+k+TY4CFM5ooQ0nBnzvg6s1SQ36yOoeNDT5++SR2RiOSLv\n"
    "xvcRviKFxmZEJCaOEDKNyJOuB56DPi/Z+fVGjmO+wea03KbNIaiGCpXZLoUmGv38\n"
    "sbZXQm2V0TP2ORQGgkE49Y9Y3IBbpNV9lXj9p5v//cWoaasm56ekBYdbqbe4oyAL\n"
    "l6lFhd2zi+WJN44pDfwGF/Y4QA5C5BIG+3vzxhFoYt/jmPQT2BVPi7Fp2RBgvGQq\n"
    "6jG35LWjOhSbJuMLe/0CjraZwTiXWTb2qHSihrZe68Zk6s+go/lunrotEbaGmAhY\n"
    "LcmsJWTyXnW0OMGuf1pGg+pRyrbxmRE1a6Vqe8YAsOf4vmSyrcjC8azjUeqkk+B5\n"
    "yOGBQMkKW+ESPMFgKuOXwIlCypTPRpgSabuY0MLTDXJLR27lk8QyKGOHQ+SwMj4K\n"
    "00u/I5sUKUErmgQfky3xxzlIPK1aEn8=\n"
    "-----END CERTIFICATE-----\n";




#endif

src/certificate_old.h
/*
 * IoTaaP Cloud Certificate - OLD
 * 
 * IoTaaP Team <contact@iotaap.io>
 * https://www.iotaap.io
 * 
 */

#ifndef __IOTAAP_CLOUD_CERTIFICATE_OLD_H__
#define __IOTAAP_CLOUD_CERTIFICATE_OLD_H__

const char *iotaap_mqtts_certificate_old =
    "-----BEGIN CERTIFICATE-----\n"
    "MIIDlTCCAn2gAwIBAgIUd6oLZtAfYIjLOg2908R3hxQZ3lUwDQYJKoZIhvcNAQEL\n"
    "BQAwWjELMAkGA1UEBhMCSFIxEDAOBgNVBAgMB0Nyb2F0aWExDzANBgNVBAcMBlph\n"
    "Z3JlYjEPMA0GA1UECgwGSW9UYWFQMRcwFQYDVQQDDA5tcXR0LmlvdGFhcC5pbzAe\n"
    "Fw0yMDAzMDgxMTA3MzhaFw0yNTAzMDcxMTA3MzhaMFoxCzAJBgNVBAYTAkhSMRAw\n"
    "DgYDVQQIDAdDcm9hdGlhMQ8wDQYDVQQHDAZaYWdyZWIxDzANBgNVBAoMBklvVGFh\n"
    "UDEXMBUGA1UEAwwObXF0dC5pb3RhYXAuaW8wggEiMA0GCSqGSIb3DQEBAQUAA4IB\n"
    "DwAwggEKAoIBAQCtKghvPBU2xnOXsx0W/5y1ew8AYuws5cSkzukgzTkQI0XTO0n5\n"
    "IT8CQD86Qi5/Q5qQtw0BQMUi5DEAPv/pHI0KuAdOaWYqlzA7AguP/4ubHvyJCNve\n"
    "Id6584YGLLw3YRSHAAFuMSwUd09NcLi21CwwSnH8K6GYhSDzYC/N9E6vzlpmSb8f\n"
    "48FeFkawHMrBgA/OFN7ffmIXe3QoCl4Ui3czk/rG3GN4J0rwpt16H8nvbGymIjBp\n"
    "9cmya8US9lFDbpuCDaKTlH7Dla6ibMkDpOA8JnwukJEwiaEJR5wN9n3NwY/JEVXa\n"
    "SXyEtQTKAqPoZj1y3CSmq/Y8FfPz1IkwPzMNAgMBAAGjUzBRMB0GA1UdDgQWBBSZ\n"
    "m4E30lB6RHa1slkKEmFUMbZFWDAfBgNVHSMEGDAWgBSZm4E30lB6RHa1slkKEmFU\n"
    "MbZFWDAPBgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQB7fWkOQ1Uj\n"
    "4/W4LLkFCHa+OYZIy+t3r8bgYjKoCqH8+DlXZWpwThjn8a+heTeYCknbh/D8270k\n"
    "edzBhUjK06qoDWpXOI4eL3+7vffLS+AEanAi/XuiqYi7spoMSz2JwXFjc/94HpYf\n"
    "WC4fcelqIdQl0qR4dnRH9otKfz4KeqhgVcjL0nyPpt2Q9UndN0nT043bthW5kk35\n"
    "f2eoPel0okXFNmcvusS1pEvg8ddfM6GkJTYZr5aj5ZjitRE0ZwGBayE76JkLrSlZ\n"
    "fbPnrsnBM5OA7ytgR7LyFB4bHaJCM81NAYDQGG9hcsiaZeCEOQHIR97xd32a52Tm\n"
    "QySj9+LAkqon\n"
    "-----END CERTIFICATE-----\n";

#endif

src/IoTaaP_HAPI.cpp
#include "IoTaaP_HAPI.h"
#include "certificate_new.h"

/**
 * @brief Construct a new IoTaaP_HAPI:: IoTaaP_HAPI object
 * 
 */
IoTaaP_HAPI::IoTaaP_HAPI(const char *fwVersion)
{
    this->_uptime = 0;
    this->_uptimePrev = 0;
    this->_sentMessages = 0;
    this->_receivedMessages = 0;
    this->_disconnects = 0;
    this->_otaUpdateNow = 0;
    this->_otaUpdatePrev = 0;
    this->_fwVersion = fwVersion;
    this->_updatesEnabled = true;
}

/**
 * @brief Configure IoTaaP Cloud connection
 * 
 * @param deviceID - Device ID from IoTaaP Cloud
 * @param deviceToken - Device Token from IoTaaP Cloud
 * @param mqttServer - IoTaaP Cloud MQTT server instance
 * @param mqttUsername - MQTT Username from IoTaaP Cloud
 * @param mqttPassword - MQTT Password from IoTaaP Cloud
 * @param groupID - (optional) Group ID from IoTaaP Cloud
 * @param groupToken - (optional) Group Token from IoTaaP Cloud
 * @return int - Returns 0 if successfull
 */
int IoTaaP_HAPI::configure(const char *deviceID, const char *deviceToken, const char *mqttServer, const char *mqttUsername, const char *mqttPassword, MQTT_CALLBACK_SIGNATURE, const char *groupID, const char *groupToken)
{
    this->_deviceID = deviceID;
    this->_deviceToken = deviceToken;
    this->_mqttServer = mqttServer;
    this->_mqttUsername = mqttUsername;
    this->_mqttPassword = mqttPassword;
    this->_groupID = groupID;
    this->_groupToken = groupToken;

    this->connectToCloud(this->_mqttServer, this->_mqttUsername, this->_mqttPassword, callback, this->_deviceID);

    return 0;
}

/**
 * @brief Execute MQTT (cloud) connection
 * 
 * @param server - MQTT server URL
 * @param user - MQTT Username
 * @param password - MQTT Password
 * @param clientID - Client ID
 * @return int - Returns 0 if successfull
 */
int IoTaaP_HAPI::connectToCloud(const char *server, const char *user, const char *password, MQTT_CALLBACK_SIGNATURE, const char *clientID)
{
    this->iotaapCore.mqtt.connect(clientID, server, 8883, callback, true, user, password, iotaap_certificate);

    return 0;
}

DynamicJsonDocument doc(256); // Dynamic JSON doucmnet used for device Status

/**
 * @brief Publishes device status periodically
 * 
 * @return int Returns 0 if successfull
 */
int IoTaaP_HAPI::publishStatus()
{

    // Uptime
    this->_uptime = millis(); // TODO - Implement overflow detection
    //

    doc["battery"] = this->iotaapCore.misc.getBatteryPercentage();
    doc["uptime"] = this->_uptime / 1000;
    doc["api_version"] = HAPI_VERSION;
    doc["fw_version"] = this->_fwVersion;

    char deviceStatus[256];

    serializeJson(doc, deviceStatus);

    String topic = "/" + String(this->_mqttUsername) + "/devices/" + String(this->_deviceID) + "/status";
    char topicChar[256];
    topic.toCharArray(topicChar, sizeof(topicChar));

    if ((this->_uptime - this->_uptimePrev) >= DEVICE_STATUS_PERIOD)
    {
        this->_uptimePrev = this->_uptime;
        this->iotaapCore.mqtt.publish(topicChar, deviceStatus, false);
    }

    return 0;
}

/**
 * @brief Publishes payload to the device topic
 * 
 * @param payload - Payload (recomended: JSON)
 * @param uTopic - Topic to publish to
 * @return int Returns 0 if successfull
 */
int IoTaaP_HAPI::devicePublish(const char *payload, const char *uTopic)
{
    String topic = "/" + String(this->_mqttUsername) + "/devices/" + String(this->_deviceID) + "/" + String(uTopic);
    char topicChar[256];
    topic.toCharArray(topicChar, sizeof(topicChar));

    this->iotaapCore.mqtt.publish(topicChar, payload, false);

    return 0;
}

/**
 * @brief Pulishes payload to the topic. Root topic (username) will be automatically added
 * 
 * @param payload - Payload (recomended: JSON)
 * @param uTopic - Topic to publish to
 * @return int - Returns 0 if successfull
 */
int IoTaaP_HAPI::publish(const char *payload, const char *uTopic)
{
    String topic = "/" + String(this->_mqttUsername) + "/" + String(uTopic);
    char topicChar[256];
    topic.toCharArray(topicChar, sizeof(topicChar));

    this->iotaapCore.mqtt.publish(topicChar, payload, false);

    return 0;
}

/**
 * @brief Subsribe to a specific topic. Root topic (username) will be added automatically
 * 
 * @param uTopic - Topic to subscribe to
 * @return int Returns 0 if successfull
 */
int IoTaaP_HAPI::subscribe(const char *uTopic)
{
    String topic = "/" + String(this->_mqttUsername) + String(uTopic);
    char topicChar[256];
    topic.toCharArray(topicChar, sizeof(topicChar));

    this->iotaapCore.mqtt.subscribe(topicChar);

    return 0;
}

/**
 * @brief Unsubscribe from the specific topic
 * 
 * @param uTopic Topic (without root)
 * @return int Returns 0 if successfull
 */
int IoTaaP_HAPI::unsubscribe(const char *uTopic)
{
    String topic = "/" + String(this->_mqttUsername) + String(uTopic);
    char topicChar[256];
    topic.toCharArray(topicChar, sizeof(topicChar));

    this->iotaapCore.mqtt.unsubscribe(topicChar);

    return 0;
}

/**
 * @brief Enables automatic updates (enabled by default)
 * 
 */
void IoTaaP_HAPI::enableUpdates(){
    this->_updatesEnabled = true;
}

/**
 * @brief Disables automatic updates (enabled by default)
 * 
 */
void IoTaaP_HAPI::disableUpdates(){
    this->_updatesEnabled = false;
}

/**
 * @brief API function that will publish IoTaaP input states to the topic 'api/transfer'. Almost non-blocking.
 * 
 */
void IoTaaP_HAPI::apiLoop(bool sendStates)
{
    this->iotaapCore.mqtt.keepAlive();
    this->publishStatus();

    this->_otaUpdateNow = millis();
    if (((this->_otaUpdateNow - this->_otaUpdatePrev) > DEVICE_OTA_CHECK_PERIOD) && this->_updatesEnabled)
    {
        this->_otaUpdatePrev = this->_otaUpdateNow;
        this->checkUpdate();
    }

    if (sendStates)
    {
        DynamicJsonDocument _doc(1024);

        _doc["device"] = this->_deviceID;

        _doc["data"]["onboard"]["but1"] = this->iotaapCore.misc.getBUT1();
        _doc["data"]["onboard"]["but2"] = this->iotaapCore.misc.getBUT2();

        _doc["data"]["digital"]["2"] = this->iotaapCore.misc.getPin(2);
        _doc["data"]["digital"]["4"] = this->iotaapCore.misc.getPin(4);
        _doc["data"]["digital"]["12"] = this->iotaapCore.misc.getPin(12);
        _doc["data"]["digital"]["13"] = this->iotaapCore.misc.getPin(13);
        _doc["data"]["digital"]["14"] = this->iotaapCore.misc.getPin(14);
        _doc["data"]["digital"]["15"] = this->iotaapCore.misc.getPin(15);
        _doc["data"]["digital"]["16"] = this->iotaapCore.misc.getPin(16);
        _doc["data"]["digital"]["17"] = this->iotaapCore.misc.getPin(17);
        _doc["data"]["digital"]["26"] = this->iotaapCore.misc.getPin(26);
        _doc["data"]["digital"]["27"] = this->iotaapCore.misc.getPin(27);
        _doc["data"]["digital"]["32"] = this->iotaapCore.misc.getPin(32);
        _doc["data"]["digital"]["33"] = this->iotaapCore.misc.getPin(33);
        _doc["data"]["digital"]["34"] = this->iotaapCore.misc.getPin(34);
        _doc["data"]["digital"]["35"] = this->iotaapCore.misc.getPin(35);

        _doc["data"]["analog"]["16"] = this->iotaapCore.misc.adc.getValue(16);
        _doc["data"]["analog"]["17"] = this->iotaapCore.misc.adc.getValue(17);
        _doc["data"]["analog"]["32"] = this->iotaapCore.misc.adc.getValue(32);
        _doc["data"]["analog"]["33"] = this->iotaapCore.misc.adc.getValue(33);
        _doc["data"]["analog"]["34"] = this->iotaapCore.misc.adc.getValue(34);
        _doc["data"]["analog"]["35"] = this->iotaapCore.misc.adc.getValue(35);

        char payload[1024];

        serializeJson(_doc, payload);

        String topic = "/" + String(this->_mqttUsername) + "/devices/" + String(this->_deviceID) + "/states";
        char topicChar[256];
        topic.toCharArray(topicChar, sizeof(topicChar));

        this->iotaapCore.mqtt.publish(topicChar, payload, false);
    }
}

/**
 * @brief Inner function to be used in MQTT callback as direct API listener on topic 'api/listen/<control_topic>'.
 * 
 * @param topic Callback topic parameter
 * @param message Callback message parameter
 * @param length Callback length parameter
 */
void IoTaaP_HAPI::callbackInnerFunction(char *topic, byte *message, unsigned int length)
{
    String messageTemp;
    DynamicJsonDocument inputBuffer(1024);

    for (int i = 0; i < length; i++)
    {
        messageTemp += (char)message[i];
    }

    String listenTopic = "/" + String(this->_mqttUsername) + "/devices/" + String(this->_deviceID) + "/listen";

    // Check if data is received on state layer
    if (String(topic) == listenTopic)
    {
        deserializeJson(inputBuffer, messageTemp);

        // Check if data type is output
        if (inputBuffer.containsKey("output"))
        {
            // Make pin as output
            this->iotaapCore.pwm.disablePWM((int)inputBuffer["output"]["pin"]);
            this->iotaapCore.misc.makePinOutput((int)inputBuffer["output"]["pin"]);
            // Check if pin is digital
            if (inputBuffer["output"]["type"] == "digital")
            {
                // Set or Clear the pin
                (int)inputBuffer["output"]["value"] ? this->iotaapCore.misc.setPin((int)inputBuffer["output"]["pin"]) : this->iotaapCore.misc.clearPin((int)inputBuffer["output"]["pin"]);
            }

            // Check if pin is analog
            if (inputBuffer["output"]["type"] == "analog")
            {
                // Set Analog value
                this->iotaapCore.pwm.setup(1, 5000, 16, (int)inputBuffer["output"]["pin"]);
                this->iotaapCore.pwm.set(1, (int)inputBuffer["output"]["value"]);
            }
        }
    }
}

/**
 * @brief Sets internal clock for CA certificate verification
 * 
 */
void IoTaaP_HAPI::setClock()
{
    configTime(0, 0, "pool.ntp.org", "time.nist.gov"); // UTC

    time_t now = time(nullptr);
    while (now < 8 * 3600 * 2)
    {
        yield();
        delay(500);
        now = time(nullptr);
    }
    struct tm timeinfo;
    gmtime_r(&now, &timeinfo);
}

/**
 * @brief Checks if new firmware version is available on the server
 * If new version is different then the current OTA update will be triggered. Function is called
 * periodically if updates are enabled.
 * 
 */
void IoTaaP_HAPI::checkUpdate()
{
    DynamicJsonDocument versionJson(128);
    this->_client.setCACert(iotaap_certificate);
    this->_client.setTimeout(12000 / 1000); // timeout argument is defined in seconds for setTimeout
    if (this->_groupID == '\0')
    { // Checking either this device is part of a gorup or standalone
        this->_httpClient.begin("https://ota.iotaap.io/v1/ota/device/latest/" + String(this->_deviceID), iotaap_certificate);
    }
    else
    {
        this->_httpClient.begin("https://ota.iotaap.io/v1/ota/group/latest/" + String(this->_groupID), iotaap_certificate);
    }

    int httpCode = this->_httpClient.GET();

    if (httpCode == 200)
    {
        String payload = this->_httpClient.getString();
        deserializeJson(versionJson, payload);

        String version = versionJson["ver"];

        if (version != this->_fwVersion)
        {
            this->otaUpdate();
        }
        else
        {
            this->_httpClient.end();
            return;
        }
    }
    else
    {
        this->_httpClient.end();
        return;
    }
}

/**
 * @brief Runs OTA Update and restarts MCU if successfull
 * 
 */
void IoTaaP_HAPI::otaUpdate()
{
    httpUpdate.setLedPin(ONBOARD_LED1, HIGH);
    httpUpdate.rebootOnUpdate(true);
    if (this->_groupID == '\0')
    {
        t_httpUpdate_return ret = httpUpdate.update(this->_client, "https://ota.iotaap.io/v1/ota/device/download/" + String(this->_deviceID) + String(this->_deviceToken));
    }
    else
    {
        t_httpUpdate_return ret = httpUpdate.update(this->_client, "https://ota.iotaap.io/v1/ota/group/download/" + String(this->_groupID) + String(this->_groupToken));
    }

    /*     switch (ret)
    {
    case HTTP_UPDATE_FAILED:
        Serial.printf("HTTP_UPDATE_FAILED Error (%d): %s\n", httpUpdate.getLastError(), httpUpdate.getLastErrorString().c_str());
        break;

    case HTTP_UPDATE_NO_UPDATES:
        Serial.println("HTTP_UPDATE_NO_UPDATES");
        break;

    case HTTP_UPDATE_OK:
        Serial.println("HTTP_UPDATE_OK");
        break;
    } */
}